### PR TITLE
Adventures

### PR DESCRIPTION
**Pull request by grena**
This PR finishes the work began by @skeleton in #47 to introduce Adventure (#40).

## Adventure list
list

## View of an adventure
step

## Progression
progression

### CODE REVIEW

File: src/Badger/GameBundle/Resources/config/doctrine/AdventureStep.orm.yml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 Badger\GameBundle\Entity\AdventureStep:
     type: entity
-    table: step
+    table: adventure_steps
```
Comment (skeleton) [inline src/Badger/GameBundle/Resources/config/doctrine/AdventureStep.orm.yml]: arf, sorry completely forgot it :(

Comment (grena) [inline src/Badger/GameBundle/Resources/config/doctrine/AdventureStep.orm.yml]: No problem, I missed it in the review too :dancer:

File: src/Badger/GameBundle/Entity/Adventure.php [hunk 2]
```diff
@@ -177,5 +181,59 @@ public function getSteps()
...
+
+    /**
+     * @return int
+     */
+    public function countAllBadges()
+    {
+        $badgesCount = (null === $this->getBadge()) ? 0 : 1;
+
+        foreach ($this->steps as $step) {
+            $badgesCount += (null ===$step->getBadge()) ? 0 : 1;
```
Comment (grena) [inline src/Badger/GameBundle/Entity/Adventure.php]: Space

File: src/Badger/GameBundle/Repository/AdventureStepCompletionRepositoryInterface.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    /**
+     * Return a collection of completed adventures steps for the given $user,
+     * grouped by adventure.
+     *
+     * @param UserInterface      $user
+     * @param AdventureInterface $adventure
+     *
+     * @return ArrayCollection
+     */
+    public function userCompletedSteps(UserInterface $user, AdventureInterface $adventure);
```
Comment (grena) [inline src/Badger/GameBundle/Repository/AdventureStepCompletionRepositoryInterface.php]: Rename to `userAdventureCompletedSteps`

File: src/Badger/GameBundle/Repository/AdventureStepCompletionRepositoryInterface.php [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    /**
+     * Return a collection of completed adventures steps for the given $user,
+     * grouped by adventure.
+     *
+     * @param UserInterface      $user
+     * @param AdventureInterface $adventure
+     *
+     * @return ArrayCollection
+     */
+    public function userClaimedSteps(UserInterface $user, AdventureInterface $adventure);
```
Comment (grena) [inline src/Badger/GameBundle/Repository/AdventureStepCompletionRepositoryInterface.php]: `userAdventureClaimedSteps`

File: src/Badger/GameBundle/Repository/AdventureStepCompletionRepositoryInterface.php [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+    /**
+     * Return informations about number of completed steps for the given $user
+     * per adventures.
+     *
+     * @param UserInterface $user
+     *
+     * @return ArrayCollection
+     */
+    public function userCompletedStepsByAdventure(UserInterface $user);
```
Comment (grena) [inline src/Badger/GameBundle/Repository/AdventureStepCompletionRepositoryInterface.php]: `userCompletedSteps`

File: src/Badger/GameBundle/Repository/UnlockedBadgeRepositoryInterface.php [hunk 6]
```diff
@@ -24,6 +25,16 @@
      */
     public function getUnlockedBadgeIdsByUser(User $user);
 
+    /**
+     * Return whether the given $user has the given $badge.
+     *
+     * @param User           $user
+     * @param BadgeInterface $badge
+     *
+     * @return mixed
```
Comment (grena) [inline src/Badger/GameBundle/Repository/UnlockedBadgeRepositoryInterface.php]: return boolean

File: src/Badger/GameBundle/Resources/config/doctrine/AdventureStepCompletion.orm.yml [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+            cascade: ['persist']
+            joinColumn:
+                user_id:
+                    referencedColumnName: id
+        step:
+            targetEntity: Badger\GameBundle\Entity\AdventureStep
+            inversedBy: completions
+            joinColumn:
+                step_id:
+                    referencedColumnName: id
```
Comment (grena) [inline src/Badger/GameBundle/Resources/config/doctrine/AdventureStepCompletion.orm.yml]: add unique constraint [ user_id, step_id ]

File: src/Badger/GateBundle/Resources/public/js/plugins/visualization/d3/d3_tooltip.js [hunk 8]
```diff
@@ -0,0 +1,328 @@
+// d3.tip
```
Comment (grena) [inline src/Badger/GateBundle/Resources/public/js/plugins/visualization/d3/d3_tooltip.js]: try without this file

File: src/Badger/GateBundle/Resources/views/adventures.html.twig [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+                    </div>
+                </div>
+            </div>
+        </div>
+    </div>
+    <!-- /grid -->
+{% endblock %}
+
+{% block javascripts %}
+{% endblock %}
```
Comment (grena) [inline src/Badger/GateBundle/Resources/views/adventures.html.twig]: remove

File: src/Badger/GateBundle/Resources/views/view-adventure.html.twig [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+                $(this).html('<i class="icon-checkmark"></i> Claimed').prop('disabled', true);
+            }.bind(this), function () {
+                l.stop();
+                $(this).html('<i class="icon-warning"></i> An error occured, click again to retry');
+            }.bind(this));
+        });
+
+
+        // Animated progress chart
+        // ------------------------------
```
Comment (grena) [inline src/Badger/GateBundle/Resources/views/view-adventure.html.twig]: Put d3 stuff in a dedicated file

File: src/Badger/GameBundle/Controller/AdventureStepCompletionController.php [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+        $user = $pendingCompletion->getUser();
+        $step = $pendingCompletion->getAdventureStep();
+        $adventure = $step->getAdventure();
+
+        $pendingCompletion->setPending(false);
+        $pendingCompletion->setCompletionDate(new \DateTime());
+
+        $errors = $this->get('validator')->validate($pendingCompletion);
+
+        if (0 === count($errors)) {
```
Comment (solivier) [inline src/Badger/GameBundle/Controller/AdventureStepCompletionController.php]: maybe we could extract the logic somewhere ? maybe create a ticket and extract it later ?

File: src/Badger/GameBundle/Doctrine/Repository/AdventureRepository.php [hunk 12]
```diff
@@ -13,4 +18,54 @@
...
+            ->andWhere()
+            ->setMaxResults(15)
+            ->groupBy('adventure.id')
+            ->setParameter('tagIds', $tagIds, Connection::PARAM_STR_ARRAY)
+        ;
+
+        return $qb->getQuery()->getResult();
+    }
+
+    public function getCompletedAdventuresForUser(TaggableInterface $user)
```
Comment (solivier) [inline src/Badger/GameBundle/Doctrine/Repository/AdventureRepository.php]: missing doc

File: src/Badger/GameBundle/Doctrine/Repository/AdventureRepository.php [hunk 13]
```diff
@@ -13,4 +18,54 @@
  */
 class AdventureRepository extends EntityRepository
```
Comment (solivier) [inline src/Badger/GameBundle/Doctrine/Repository/AdventureRepository.php]: interface ?

File: src/Badger/GameBundle/Unlocker/BadgeUnlocker.php [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+use Badger\StorageUtilsBundle\Remover\RemoverInterface;
+use Badger\StorageUtilsBundle\Saver\SaverInterface;
+use Badger\UserBundle\Entity\UserInterface;
+
+/**
+ * @author    Adrien Pétremann <adrien.petremann@akeneo.com>
+ * @copyright 2016 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class BadgeUnlocker
```
Comment (solivier) [inline src/Badger/GameBundle/Unlocker/BadgeUnlocker.php]: we should create an unlocker interface maybe ?

File: src/Badger/GameBundle/Controller/AdventureStepCompletionController.php (hunk 11 above)
Comment (grena) [inline src/Badger/GameBundle/Controller/AdventureStepCompletionController.php]: Yep you're right. It bothers me too to have this kind of logic in a Controller.
I don't know if we should use events or just kind of "BadgeUnlocker" but for Quest & AdventureSteps too.

File: src/Badger/GameBundle/Doctrine/Repository/AdventureRepository.php (hunk 13 above)
Comment (grena) [inline src/Badger/GameBundle/Doctrine/Repository/AdventureRepository.php]: done

File: src/Badger/GameBundle/Doctrine/Repository/AdventureRepository.php (hunk 12 above)
Comment (grena) [inline src/Badger/GameBundle/Doctrine/Repository/AdventureRepository.php]: done

File: src/Badger/GameBundle/Unlocker/BadgeUnlocker.php (hunk 14 above)
Comment (grena) [inline src/Badger/GameBundle/Unlocker/BadgeUnlocker.php]: done

File: src/Badger/GameBundle/Controller/AdventureStepCompletionController.php (hunk 11 above)
Comment (grena) [inline src/Badger/GameBundle/Controller/AdventureStepCompletionController.php]: I've created this https://github.com/akeneo/badger/issues/113

File: src/Badger/GateBundle/Controller/DefaultController.php [hunk 15]
```diff
@@ -178,6 +177,42 @@ public function claimBadgeAction($id)
...
+    {
+        $user = $this->getUser();
+        $step = $this->get('badger.game.repository.adventure_step')->find($id);
+
+        if (null === $step) {
+            return new JsonResponse('No step with this id.', 400);
+        }
+
+        if (!$this->get('security.authorization_checker')->isGranted('view', $step->getAdventure())) {
+            return new JsonResponse('No step with this id.', 400);
```
Comment (skeleton) [inline src/Badger/GateBundle/Controller/DefaultController.php]: 401 ?

File: src/Badger/GameBundle/Doctrine/Repository/AdventureRepository.php [hunk 16]
```diff
@@ -11,6 +14,55 @@
...
+            $tagIds[] = $tag->getId();
+        }
+
+        $qb = $this->createQueryBuilder('adventure');
+        $qb->leftJoin('adventure.tags', 'tags')
+            ->leftJoin('adventure.steps', 's')
+            ->leftJoin('s.completions', 'sc')
+            ->having('COUNT(sc) < COUNT(s)')
+            ->andWhere('tags.id IN (:tagIds)')
+            ->andWhere()
```
Comment (grena) [inline src/Badger/GameBundle/Doctrine/Repository/AdventureRepository.php]: done

File: src/Badger/GameBundle/Doctrine/Repository/AdventureRepository.php [hunk 17]
```diff
@@ -11,6 +14,55 @@
...
+            $tagIds[] = $tag->getId();
+        }
+
+        $qb = $this->createQueryBuilder('adventure');
+        $qb->leftJoin('adventure.tags', 'tags')
+            ->leftJoin('adventure.steps', 's')
+            ->leftJoin('s.completions', 'sc')
+            ->having('COUNT(sc) = COUNT(s)')
+            ->andWhere('tags.id IN (:tagIds)')
+            ->andWhere()
```
Comment (grena) [inline src/Badger/GameBundle/Doctrine/Repository/AdventureRepository.php]: done

File: src/Badger/GameBundle/Entity/AdventureInterface.php [hunk 18]
```diff
@@ -96,7 +97,7 @@ public function setBadge(Badge $badge);
     /**
      * Get badge
      *
-     * @return Badge
+     * @return Badge|null
```
Comment (grena) [inline src/Badger/GameBundle/Entity/AdventureInterface.php]: done

File: src/Badger/GameBundle/Form/AdventureType.php [hunk 19]
```diff
@@ -43,6 +43,13 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 'by_reference' => false,
                 'label'        => false
             ])
+            ->add('tags', 'entity', [
```
Comment (grena) [inline src/Badger/GameBundle/Form/AdventureType.php]: done

File: src/Badger/GameBundle/Form/AdventureType.php [hunk 20]
```diff
@@ -43,6 +43,13 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
+                'label'    => 'Tagged in',
+                'multiple' => true,
+                'property' => 'name',
```
Comment (grena) [inline src/Badger/GameBundle/Form/AdventureType.php]: done

File: src/Badger/GameBundle/Repository/AdventureStepCompletionRepositoryInterface.php [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+interface AdventureStepCompletionRepositoryInterface
+{
+    /**
+     * Return a collection of completed adventures steps for the given $user,
+     * grouped by adventure.
+     *
+     * @param UserInterface      $user
+     * @param AdventureInterface $adventure
+     *
+     * @return ArrayCollection
```
Comment (grena) [inline src/Badger/GameBundle/Repository/AdventureStepCompletionRepositoryInterface.php]: done

File: src/Badger/GameBundle/Repository/AdventureStepCompletionRepositoryInterface.php [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+    public function userAdventureCompletedSteps(UserInterface $user, AdventureInterface $adventure);
+
+    /**
+     * Return a collection of completed adventures steps for the given $user,
+     * grouped by adventure.
+     *
+     * @param UserInterface      $user
+     * @param AdventureInterface $adventure
+     *
+     * @return ArrayCollection
```
Comment (grena) [inline src/Badger/GameBundle/Repository/AdventureStepCompletionRepositoryInterface.php]: done

File: src/Badger/GameBundle/Repository/AdventureStepCompletionRepositoryInterface.php [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+     */
+    public function userAdventureClaimedSteps(UserInterface $user, AdventureInterface $adventure);
+
+    /**
+     * Return informations about number of completed steps for the given $user
+     * per adventures.
+     *
+     * @param UserInterface $user
+     *
+     * @return ArrayCollection
```
Comment (grena) [inline src/Badger/GameBundle/Repository/AdventureStepCompletionRepositoryInterface.php]: done

File: src/Badger/GameBundle/Repository/UnlockedBadgeRepositoryInterface.php [hunk 24]
```diff
@@ -24,6 +25,16 @@
...
 
+    /**
+     * Return whether the given $user has the given $badge.
+     *
+     * @param User           $user
+     * @param BadgeInterface $badge
+     *
+     * @return boolean
+     */
+    public function userHasBadge(User $user, BadgeInterface $badge);
```
Comment (grena) [inline src/Badger/GameBundle/Repository/UnlockedBadgeRepositoryInterface.php]: done

File: src/Badger/GateBundle/Controller/DefaultController.php [hunk 25]
```diff
@@ -178,6 +177,42 @@ public function claimBadgeAction($id)
...
+     *
+     * @return JsonResponse
+     */
+    public function claimStepAction($id)
+    {
+        $user = $this->getUser();
+        $step = $this->get('badger.game.repository.adventure_step')->find($id);
+
+        if (null === $step) {
+            return new JsonResponse('No step with this id.', 400);
```
Comment (grena) [inline src/Badger/GateBundle/Controller/DefaultController.php]: done

File: src/Badger/GateBundle/Controller/DefaultController.php (hunk 15 above)
Comment (grena) [inline src/Badger/GateBundle/Controller/DefaultController.php]: done

**Pull request merged**